src/cls/rbd/cls_rbd.cc

```diff
@@ -1418,147 +1576,194 @@ int get_parent(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+  return 0;
+}
+
+
+/**
+ * remove the parent pointer
+ *
+ * This can only happen on the head, not on a snapshot.  No arguments.
+ *
+ * @returns 0 on success, negative error code on failure.
+ */
+int remove_parent(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  int r = parent::detach(hctx, true);
   if (r < 0) {
-    CLS_LOG(20, "cls_rbd::set_parent: child does not support layering");
     return r;
   }
 
-  CLS_LOG(20, "set_parent pool=%llu id=%s snapid=%llu size=%llu",
-	  (unsigned long long)pool, id.c_str(), (unsigned long long)snapid.val,
-	  (unsigned long long)size);
+  return 0;
+}
+
 
-  if (pool < 0 || id.length() == 0 || snapid == CEPH_NOSNAP || size == 0) {
-    return -EINVAL;
+/**
+ * Input:
+ * none
+ *
+ * Output:
+ * @param parent spec (cls::rbd::ParentImageSpec)
+ * @returns 0 on success, negative error code on failure
+ */
+int parent_get(cls_method_context_t hctx, bufferlist *in, bufferlist *out) {
+  int r = check_exists(hctx);
+  if (r < 0) {
+    return r;
   }
 
-  // make sure there isn't already a parent
+  CLS_LOG(20, "parent_get");
+
   cls_rbd_parent parent;
-  r = read_key(hctx, "parent", &parent);
+  r = require_feature(hctx, RBD_FEATURE_LAYERING);
   if (r == 0) {
-    CLS_LOG(20, "set_parent existing parent pool=%llu id=%s snapid=%llu"
-            "overlap=%llu", (unsigned long long)parent.pool_id,
-            parent.image_id.c_str(),
-	    (unsigned long long)parent.snap_id.val,
-	    (unsigned long long)parent.overlap);
-    return -EEXIST;
-  }
-
-  // our overlap is the min of our size and the parent's size.
-  uint64_t our_size;
-  r = read_key(hctx, "size", &our_size);
-  if (r < 0)
-    return r;
+    r = read_key(hctx, "parent", &parent);
+    if (r < 0 && r != -ENOENT) {
+      return r;
+    } else if (r == -ENOENT) {
+      // examine oldest snapshot to see if it has a denormalized parent
+      auto parent_lambda = [hctx, &parent](const cls_rbd_snap& snap_meta) {
+        if (snap_meta.parent.exists()) {
+          parent = snap_meta.parent;
+        }
+        return 0;
+      };
 
-  bufferlist parentbl;
-  parent.pool_id = pool;
-  parent.image_id = id;
-  parent.snap_id = snapid;
-  parent.overlap = std::min(our_size, size);
-  encode(parent, parentbl);
-  r = cls_cxx_map_set_val(hctx, "parent", &parentbl);
-  if (r < 0) {
-    CLS_ERR("error writing parent: %s", cpp_strerror(r).c_str());
-    return r;
+      r = image::snapshot_iterate(hctx, parent_lambda);
+      if (r < 0) {
+        return r;
+      }
+    }
   }
 
+  cls::rbd::ParentImageSpec parent_image_spec{
+    static_cast<cls::rbd::ParentImageSpec>(parent)};
+  encode(parent_image_spec, *out);
   return 0;
 }
 
-
 /**
- * remove the parent pointer
- *
- * This can only happen on the head, not on a snapshot.  No arguments.
+ * Input:
+ * @param snap id (uint64_t) parent snapshot id
  *
- * @returns 0 on success, negative error code on failure.
+ * Output:
+ * @param byte overlap of parent image (uint64_t)
```

Comment:
@idryomov Is it too late for me to change this to an ```std::optional<uint64_t>```? It would just add a bool byte to the front to indicate whether or not the overlap exists or not. I think I discovered a backwards compatibility issue by trying to normalize the parent image spec (by removing the duplication from the snapshots) that prevents librbd from detecting whether or not the link exists on-disk when the overlap is actually zero (i.e. clone a zero byte parent or shrink a clone down to zero bytes).

Reply:
No, it's not too late.  The patch that implements parent_get/parent_overlap_get isn't upstream yet.  Worst case, the updated patch doesn't get merged, but we still have the "if the client comes in through the old get_parent method, return an error if the parent image is in a custom namespace" protection.